File: Leetcode100/leetcode100_python/ShortestUnsortedContinuousSubarray2.py
from typing import List

class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        if n <= 1:
            return 0

        left, right = n, 0
        stack = []
        for i in range(n):
            while stack and nums[i] < nums[stack[-1]]:
                left = min(left, stack.pop())
            stack.append(i)

        stack = []
        for i in range(n - 1, -1, -1):
            while stack and nums[i] > nums[stack[-1]]:
                right = max(right, stack.pop())
            stack.append(i)

        return 0 if right - left < 0 else right - left + 1

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findUnsortedSubarray(nums))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/shortest_unsorted_continuous_subarray/ShortestUnsortedContinuousSubarray2.java
package main.java.com.jiang.leetcode100.shortest_unsorted_continuous_subarray;

import java.util.Arrays;
import java.util.Scanner;

class Solution2 {
    public int findUnsortedSubarray(int[] nums) {
        int left = nums.length, right = 0;
        for(int i = 0; i < nums.length; i++) {
            for(int j = i + 1; j < nums.length; j++) {
                if(nums[j] < nums[i]) {
                    right = Math.max(right, j);
                    left = Math.min(left, i);
                }
            }
        }
        return right - left < 0 ? 0 : right - left + 1;
    }
}

public class ShortestUnsortedContinuousSubarray2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        System.out.println(new Solution2().findUnsortedSubarray(nums));
    }
}